File: java/1985-find-the-kth-largest-integer-in-the-array.java
// Min Heap comparator for reverse priority queue
class StringNumberComparartor implements Comparator<String> {
    @Override
    public int compare(String s1, String s2) {
        if(s1.length() != s2.length()) {
            return (s1.length() - s2.length());
        }
        int len = s1.length();
        for(int i=0; i<len; i++) {
            char c1 = s1.charAt(i), c2 = s2.charAt(i);
            if(c1 == c2) { continue; }
            return (c1 - c2);
        }
        return 0;
    }
}

class Solution {
    public String kthLargestNumber(String[] nums, int k) {
        // Reverse pq to store only k elements [i.e. Min Heap]
        PriorityQueue<String> pq = new PriorityQueue<>(new StringNumberComparartor());

        for(String numStr: nums) {
            pq.add(numStr);
            if(pq.size() > k) {
                pq.remove();
            }
        }

        return pq.peek();
    }
}

File: kotlin/1985-find-the-kth-largest-integer-in-the-array.kt
/*
* Using a minHeap (And BigInteger, alternativly you can compare strings too)
*/
class Solution {
    fun kthLargestNumber(nums: Array<String>, k: Int): String {

        val minHeap = PriorityQueue<BigInteger>{ a, b -> 
            if(a < b)
                1
            else
                -1
        }

        nums.forEach {
            minHeap.add(it.toBigInteger())
        }

        var kth = k
        while(kth > 1) {
            minHeap.poll()
            kth--
        }

        return minHeap.poll().toString()
    }
}

/*
* Using a maxHeap (And comparing strings, alternativly you can use BigInteger too)
*/
class Solution {
    fun kthLargestNumber(nums: Array<String>, k: Int): String {

        val minHeap = PriorityQueue<String>{ a, b -> 
            if(a.length > b.length)
                1
            else if(a.length < b.length)
                -1
            else {
                var toReturn = 0
                for(i in 0 until a.length) {
                    if(a[i] > b[i]) {
                        toReturn = 1
                        break
                    } else if(a[i] < b[i]){
                        toReturn = -1
                        break
                    }
                }
                toReturn
            }
        }

        nums.forEach {
            minHeap.add(it)
            if(minHeap.size > k) minHeap.poll()
        }

        return minHeap.peek()
    }
}
